File: 376. Wiggle Subsequence.cpp
#include <vector>

class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if(nums.size() <2){
            return nums.size();
        }
        static const int begin = 0;
        static const int up = 1;
        static const int down =2;
        int state = begin;
        int max_length = 1;
        
        for (int i=1; i <nums.size(); i++){
            switch(state){
                case begin:
                    if(nums[i-1] < nums[i]){
                        state = up;
                        max_length++;
                    }
                    else if(nums[i-1] > nums[i]){
                        state = down;
                        max_length++;
                    }
                    break;
                case up:
                    if(nums[i-1] > nums[i]){
                        state = down;
                        max_length++;
                    }
                    break;
                case down:
                    if(nums[i-1] < nums[i]){
                        state = up;
                        max_length++;
                    }
                    break;
            }
        }
        return max_length;
    }
};

File: 376. Wiggle Subsequence.py
class Solution:
    def wiggleMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) <2:
            return len(nums)
        begin = 0
        up    = 1
        down  = 2
        state = begin
        max_length = 1
        for i in range(1,len(nums)):
            if state == begin:
                if nums[i-1] < nums[i]:
                    state = up
                    max_length +=1
                elif nums[i-1] > nums[i]:
                    state = down
                    max_length +=1
            if state == up:
                if nums[i-1] > nums[i]:
                    state = down
                    max_length +=1
            if state == down:
                if nums[i-1] < nums[i]:
                    state = up
                    max_length +=1    
            i+=1
        return max_length
